src/SnapKeyring.test.ts

```diff
@@ -665,6 +552,131 @@ describe('SnapKeyring', () => {
       expect(await keyring.getAccounts()).toStrictEqual([]);
     });
   });
+  describe('async request redirect', () => {
+    it.each([
+      [{ message: 'Go to dapp to continue.' }],
+      [{ url: 'https://example.com/sign?tx=1234' }],
+      [{ url: 'https://example.com/sign?tx=12345', message: 'Go to dapp.' }],
+    ])('returns a redirect %s', async (redirect) => {
+      const spy = jest.spyOn(console, 'log').mockImplementation();
+
+      const snapObject = {
+        id: snapId,
+        manifest: {
+          initialPermissions: {
+            'endowment:keyring': {
+              allowedOrigins: ['https://example.com'],
+            },
+          },
+        },
+        enabled: true,
+      };
+      mockSnapController.get.mockReturnValue(snapObject);
+
+      mockSnapController.handleRequest.mockResolvedValue({
+        pending: true,
+        redirect,
+      });
+      const requestPromise = keyring.signPersonalMessage(
+        ethEoaAccount1.address,
+        'hello',
+      );
+
+      const { calls } = mockSnapController.handleRequest.mock;
+      const requestId = calls[calls.length - 1][0].request.params.id;
+      await keyring.handleKeyringSnapMessage(snapId, {
+        method: KeyringEvent.RequestRejected,
+        params: { id: requestId },
+      });
+
+      const { url = '', message = '' } = redirect as {
+        url?: string;
+        message?: string;
+      };
+
+      // We need to await on the request promise because the request submission
+      // is async, so if we don't await, the test will exit before the promise
+      // gets resolved.
+      await expect(requestPromise).rejects.toThrow(
+        'Request rejected by user or snap.',
+      );
+
+      // Check that `redirectUser` was called with the correct parameters
+      expect(mockCallbacks.redirectUser).toHaveBeenCalledWith(
+        snapId,
+        url,
+        message,
+      );
+      spy.mockRestore();
+    });
+    const isNotAllowedOrigin = async (
+      allowedOrigins: string[],
+      redirectUrl: string,
+    ) => {
+      const { origin } = new URL(redirectUrl);
+      const snapObject = {
+        id: snapId,
+        manifest: {
+          initialPermissions:
+            allowedOrigins.length > 0
+              ? { 'endowment:keyring': { allowedOrigins } }
+              : {},
+        },
+        enabled: true,
+      };
+      mockSnapController.get.mockReturnValue(snapObject);
+      mockSnapController.handleRequest.mockResolvedValue({
+        pending: true,
+        redirect: {
+          message: 'Go to dapp to continue.',
+          url: redirectUrl,
+        },
+      });
+      const requestPromise = keyring.signPersonalMessage(
+        ethEoaAccount1.address,
+        'hello',
+      );
+
+      await expect(requestPromise).rejects.toThrow(
+        `Redirect URL domain '${origin}' is not an allowed origin by snap '${snapId}'`,
+      );
+    };
```

Comment:
Let's move this function before the `it.each` of this scope.

Reply:
good idea! Addressed here [1de55bd](https://github.com/MetaMask/eth-snap-keyring/pull/301/commits/1de55bd86b7448457da73cd6d35d72fb805bcaac) 😄